### PR TITLE
[zero] add unit testings for hybrid parallelism 

### PR DESCRIPTION
**Pull request by 1SAA**
### What's New

I added a unit testing for hybrid parallelism. It makes sure that the combination of low-level zero and tensor parallelism works fine.  Currently, the testing fails when `overlap_communication` is set to `True`. I'm fixing this bug now.

### UPDATE (1):

I found that the gradient hook can't be triggered on `ColoParameter`. Thus, I use the function `register_hook` to add a gradient hook for each parameter. I also polished the code in the zero optimizer.

- [x] fix the bug when `overlap_communication` is set to `True`.

### CODE REVIEW

File: colossalai/zero/sharded_optim/low_level_optim.py [hunk 1]
```diff
@@ -268,53 +272,61 @@ def _attach_reduction_hook(self):
...
-                    def _define_and_attach(param, reduce_rank):
-                        # get the AccumulateGrad object of the param itself
-                        accum_grad_obj = get_grad_accumulate_object(param)
-                        self._grad_store.add_accumulate_grad_object(accum_grad_obj)
+                    param.register_hook(partial(self._grad_handler, param, reduce_rank=reduce_rank))
 
-                        reduction_func = partial(self._reduce_and_remove_grads_by_bucket,
-                                                 param=param,
-                                                 reduce_rank=reduce_rank)
+    def _reduce_tensor_bucket(self, bucket: TensorBucket, reduce_rank):
```
Comment (fangjiarui) [inline colossalai/zero/sharded_optim/low_level_optim.py:277]: 如果self._overlap_communication:=True，这个接口是一个异步的
但是我没找到 wait_stream操作。

参考了Stream semantics例子
https://pytorch.org/docs/stable/notes/cuda.html#cuda-semantics

Comment (feifeibear) [inline colossalai/zero/sharded_optim/low_level_optim.py:277]: 如果self._overlap_communication:=True，这个接口是一个异步的
但是我没找到 wait_stream操作。

参考了Stream semantics例子
https://pytorch.org/docs/stable/notes/cuda.html#cuda-semantics

Comment (feifeibear) [reply colossalai/zero/sharded_optim/low_level_optim.py:277]: 我理解 with torch.cuda.stream(s): 要和一个wait_stream配合使用

Comment (1SAA) [reply colossalai/zero/sharded_optim/low_level_optim.py:277]: 其实不需要wait_stream函数，代码里面使用torch.cuda.synchronize()同步了所有的stream。这个代码在不同stream上都是同步的，但是由于GPU会管理不同stream并行，所以在默认stream上的计算和在另一个stream上的通信是异步的。

**Review by feifeibear — APPROVED** [review]

File: colossalai/zero/sharded_optim/low_level_optim.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/zero/sharded_optim/low_level_optim.py:277]: OK，你觉得没问题就行。

**Pull request merged**